Secure user data

Rick-Anderson opened the pull request:
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/security/authorization/secure-data?branch=pr-en-us-3336)

Rick-Anderson commented:
@joeaudette  I got rid of all the extra commits. Your code is in the *samples/final15* folder - right? So I can delete *samples/final* ?

joeaudette commented:
@Rick-Anderson yes, my changes were in samples/final15

Rick-Anderson commented:
@joeaudette When I create a new user rick@example.com and create a couple new contacts, another new user (joe@example.com) can see them before they are approved. Is that the intended behavior? I see that manager@contoso.com  has an approve/reject button. That button only seems to set the status field. Is that the intended behavior? I was expecting  registered user could only see approved contacts. We don’t need that as long as the code sample shows people enough to get started on writing an app with secured user data.

joeaudette commented:
I see your point, it might be more realistic to filter out the unapproved ones from the list for users who can't approve. So yes intended but maybe not as well thought out or as good an example as it could be. Really the approval was just to introduce new operations to differentiate normal users from more privileged users. I figured by showing the status but not showing a way to change the status unless logged in as a more privileged user was sufficient. Users would need to see their own unapproved contacts anyway, so if we were to filter unapproved ones we would have to filter only those that belong to others. We can change it though if you want to filter out unapproved contacts from other users. I guess I was thinking that multiple users might be adding the same person as a contact, and by showing both approved and unapproved with status, a user would know that the person is already in the list and not add a duplicate. It was intended that the approve/reject button only changes the status, if I recall correctly the manager cannot edit other user's contacts but can only approve or reject them whereas admin can edit any contacts well as approve or reject.

Rick-Anderson commented:
```
public async Task<IActionResult> Index()
{
    var contacts = from c in _context.Contact
                   select c;

    // REVIEW: Must be a cleaner way.
    var contactDB = await _context.Contact.FirstOrDefaultAsync();

    var isAuthorized = await _authorizationService.AuthorizeAsync(
                                               User, contactDB,
                                               ContactOperations.Approve);
    // Review: good if users could see their own contacts before approval.
    if (!isAuthorized)
    {
        contacts = contacts.Where(c => c.Status == ContactStatus.Approved);
    }

    return View(await contacts.ToListAsync());
}
```

{
    if (id == null)
    {
        return NotFound();
    }

    var contact = await _context.Contact.SingleOrDefaultAsync(m => m.ContactId == id);
    if (contact == null)
    {
        return NotFound();
    }

    var isAuthorized = await _authorizationService.AuthorizeAsync(
                                               User, contact,
                                               ContactOperations.Approve);

    if (contact.Status != ContactStatus.Approved && !isAuthorized)
    {
        return new ChallengeResult();
    }

    return View(contact);
}

Rick-Anderson commented:
@joeaudette  can you review the code and the article?  In the code search for `review` match whole word (not case sensitive). 

I changed  `ContactIsOwnerAuthorizationHandler`  and removed

```
if (requirement.Name == Constants.ApproveOperationName ||
    requirement.Name == Constants.RejectOperationName)
{
    return Task.FromResult(0);
}
```
that seemed to be excluding other permissions and mixing ownership logic with approve logic. When other permissions are added, we'll need to exclude them too. 

I also refactored the create manager/admin code.

tdykstra commented on aspnetcore/security/authorization/secure-data.md:
```diff
@@ -0,0 +1,314 @@
+
+## Update the code to support authorization
+
+In this section, you update the controller, views and add an operations requirements class.
```
In this section, you update the controller and views and add an operations requirements class.

tdykstra commented on aspnetcore/security/authorization/secure-data.md:
```diff
@@ -0,0 +1,314 @@
+
+[!code-csharp[Main](secure-data/samples/final/Controllers/ContactsController.cs?name=snippet_Delete)]
+
+## Inject the authorization service into the views
```
s/b h3

Rick-Anderson replied:
That's a major new thing so I'd like to keep it in the TOC.

tdykstra commented on aspnetcore/security/authorization/secure-data.md:
```diff
@@ -0,0 +1,314 @@
+
+[!code-html[Main](secure-data/samples/final/Views/_ViewImports.cshtml)]
+
+Update the *Views/Contacts/Index.cshtml* Razor view to show only display the edit and delete links for users who can edit/delete the contact.
```
to show only display --> to display ... only for user who can ...

tdykstra commented on aspnetcore/security/authorization/secure-data.md:
```diff
@@ -0,0 +1,314 @@
+
+[!code-html[Main](secure-data/samples/final/Views/Contacts/Index.cshtml?range=63-84)]
+
+Warning: Hiding links from users that do not have permission to edit or delete data does not secure that app, it makes the app more user friendly by displaying only valid links. Users can hack the generated URLs to invoke edit and delete operations on data they don't own.  The controller must repeat the access checks to be secure.
```
secure that app --> secure the app

tdykstra commented on aspnetcore/security/authorization/secure-data.md:
```diff
@@ -0,0 +1,314 @@
+
+You'll need to register a user to browse the contacts.
+
+An easy way to test the completed app is to launch three different browsers (or incognito/InPrivate versions). In one browser, register a new user, for example, `test@example.com`. SignIn to each browser with a different user. Verify the following:
```
SignIn --> Sign in

tdykstra commented on aspnetcore/security/authorization/secure-data.md:
```diff
@@ -0,0 +1,314 @@
+| manager@contoso.com | Can approve/reject and edit/delete own data  |
+| test@example.com | Can edit/delete own data |
+
+Create a contact in the administrators browser. Copy the URL for delete and edit from the administrator contact. Paste these links into the test users browser to verify the test user cannot perform these operations.
```
test users browser -> test user's browser

tdykstra commented on aspnetcore/security/authorization/secure-data.md:
```diff
@@ -0,0 +1,314 @@
+
+Follow these instructions to create the starter app.
+
+* Create a new **ASP.NET Core Web Application** using [Visual Studio 2017](https://www.visualstudio.com/en-us/visual-studio-homepage-vs.aspx) named "ContactManager"
```
These instructions say use VS 2017 but earlier you say "if you're using VS 2017"

Rick-Anderson replied:
That's just to create the starter app from scratch, which is optional. VS Code folks will probably download the starter.

BTW, I really hate writing "to create it from scratch" but I can't think of a better way to say that.

tdykstra commented on aspnetcore/security/authorization/secure-data.md:
```diff
@@ -0,0 +1,314 @@
+
+* Test the app by creating, editing and deleting a contact
+
+## Seed the database
```
This is still creating the starter, should probably be h3

tdykstra commented on aspnetcore/security/authorization/secure-data.md:
```diff
@@ -0,0 +1,314 @@
+
+Test that the app seeded the database. The seed method does not run if there are any rows in the contact DB.
+
+### Create the resources for the tutorial
```
Why are you calling this "resources"?

tdykstra commented on aspnetcore/security/authorization/secure-data.md:
```diff
@@ -0,0 +1,314 @@
+
+### Additional resources
+
+* [ASP.NET Core Authorization Lab](https://github.com/blowdart/AspNetAuthorizationWorkshop)
```
Would be helpful to explain what they'll get there that they don't get in the tutorial.

tdykstra commented on aspnetcore/security/authorization/secure-data.md:
```diff
@@ -0,0 +1,314 @@
+### Additional resources
+
+* [ASP.NET Core Authorization Lab](https://github.com/blowdart/AspNetAuthorizationWorkshop)
+* [Authorization](index.md)
```
Authorization --> Authorization in ASP.NET Core : Simple, role, claims-based and custom

Rick-Anderson commented on aspnetcore/security/authorization/secure-data.md:
```diff
@@ -0,0 +1,314 @@
+
+In the following image, `manager@contoso.com` is signed in and in the managers role. 
+
+![image described above](secure-data/_static/manager1.png)
```
Red has really poor contrast with back.

Rick-Anderson commented on aspnetcore/security/authorization/secure-data.md:
```diff
@@ -0,0 +1,314 @@
+
+The contact information properties (Address, Name, etc.) are displayed in the images above. `ContactId` is the primary key for the table.
+
+A `ContactIsOwnerAuthorizationHandler` authorization handler ensures that a user can only edit their data. A `ContactManagerAuthorizationHandler` authorization handler allows managers to approve or reject contacts.  A `ContactAdministratorsAuthorizationHandler` authorization handler allows administrators to approve or reject contacts and to edit/delete contacts. 
```
@joeaudette  should I rename it `ContactUserAuthorizationHandler` ? I sort of like the original as it denotes _IsOwner_. They all start with Contact.

Rick-Anderson commented:
@blowdart  can you give it one last blessing? [Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/security/authorization/secure-data?branch=pr-en-us-3336)

joeaudette commented:
@Rick-Anderson I reviewed as requested. A couple of suggestions.

1. We can remove this from Startup.cs since we already have global filter to require authenticated user:

     services.AddAuthorization(options =>
            {
                options.AddPolicy(
                Constants.ContactUserPolicy,
                authBuilder =>
                {
                    authBuilder.RequireAuthenticatedUser();
                });
            });

and remove the Constant public static readonly string ContactUserPolicy = "ContactUserPolicy";

and remove the policy from ContactsController [Authorize(Policy = "ContactUserPolicy")]

2. The isAuthorized in ContactsContoller can be simplified to just check roles and to allow users to see their own unapproved contacts like this:

        public async Task<IActionResult> Index()
        {
            var contacts = from c in _context.Contact
                           select c;

            var isAuthorized = User.IsInRole(Constants.ContactManagersRole) || User.IsInRole(Constants.ContactAdministratorsRole);
            
            var currentUserId = _userManager.GetUserId(User);

            // Only approved contacts are shown UNLESS you're authorized to see them all or you are the owner.
            if (!isAuthorized)
            {
                contacts = contacts.Where(c => c.Status == ContactStatus.Approved || c.OwnerID == currentUserId);
            }

            return View(await contacts.ToListAsync());
        }

3. There is a stray project.json file the project that needs to be deleted

joeaudette commented:
@Rick-Anderson more feedback about the sample

currently we have in startup

     services.Configure<MvcOptions>(options =>
     {
         options.Filters.Add(new RequireHttpsAttribute());
     });

which does illustrate a good practice. However it makes the project not work by default because no ssl certificate is available. This adds a significant friction for users especially if they are on mac or linux using VSCode. In VS2017 we could solve it at least in IIS Expresss by adding https in launchSettings.json like this: I think IISExpress or tooling will auto generate a self signed cert.

    {
      "iisSettings": {
        "windowsAuthentication": false,
        "anonymousAuthentication": true,
        "iisExpress": {
          "applicationUrl": "http://localhost:2223/",
          "sslPort": 44366
        }
      },
       "profiles": {
        "IIS Express": {
          "commandName": "IISExpress",
          "launchBrowser": true,
          "launchUrl": "https://localhost:44366/",
          "environmentVariables": {
            "ASPNETCORE_ENVIRONMENT": "Development"
          }
       },
        "final15": {
         "commandName": "Project"
       }
     }
   }

that file is not included in git but could be added to the download?
But even with that running it from the command line directly with kestrel via dotnet run will fail in the absence of an ssl certificate being setup which is probably beyond the scope of the article but also a major barrier for anyone new trying the sample.

I would suggest changing the Startup code like this:

    var useSsl = Configuration.GetValue<bool>("AppSettings:UseSsl");
            services.Configure<MvcOptions>(options =>
            {
                if (useSsl)
                {
                    options.Filters.Add(new RequireHttpsAttribute());
                }
            });

and setting it to false in appsettings.config like this:

    "AppSettings": {
    "UseSsl": false
  }

at least then the sample works and we still illustrate how to require ssl in production

Rick-Anderson commented:
@joeaudette  awesome feedback. SSL code change - I also explain skipping SSL **twice** in the article
```
 var skipSSL = Configuration.GetValue<bool>("LocalTest:skipSSL");
 // requires using Microsoft.AspNetCore.Mvc;
 services.Configure<MvcOptions>(options =>
 {
     // Set LocalTest:skipSSL to true to skip SSL requrement in 
     // debug mode. This is useful when not using Visual Studio.
     if (_hostingEnv.IsDevelopment() && !skipSSL)
     {
         options.Filters.Add(new RequireHttpsAttribute());
     }
 });
```
I verified the download includes *launchSettings.json*. I did a `git -f add solution Explorer` too, so the launch settings will work.

Rick-Anderson commented:
@blowdart  can you give it a quick review - not too much has changed since you last approved it. This is our customers #1 request.
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/security/authorization/secure-data?branch=pr-en-us-3336)

Rick-Anderson commented:
@blowdart  can you give it a quick review - not too much has changed since you last approved it. This is our customers #1 request.
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/security/authorization/secure-data?branch=pr-en-us-3336)

blowdart commented:
Hah, #1 linked to an issue. That was confusing.

This is great, especially like there's different scopes for the handlers.

:shipit:
